File: lee/src/main/java/org/example/lee/题目/链表/c_K个一组翻转链表.go
package 链表

func reverseKGroup(head *ListNode, k int) *ListNode {
	//固定模板
	i := 0
	a := head
	//判断是否为k个 不是直接返回 是则获得一个下一个区间的首元素
	for i < k {
		if a == nil {
			return head
		}
		a = a.Next
		i++
	}
	i = 0
	for i < k {
		help := head.Next
		if i == 0 {
			//当i=0,让head连接下一个区间的首元素
			head.Next = reverseKGroup(a, k)
		} else {
			//此时a保存的是head翻转后应该连接的元素
			head.Next = a
		}
		a = head
		head = help
		i++
	}
	//最终的a就是头元素
	return a
}

File: lee/src/main/java/org/example/lee/题目/链表/c_k个一组翻转链表.java
package org.example.lee.题目.链表;

import org.example.lee.model.ListNode;

public class c_k个一组翻转链表 {


  /**
   * reverse kgroup 递归+倒序
   *
   * @param head head
   * @param k    k
   * @return {@link ListNode}
   */
  public static ListNode reverseKGroup(ListNode head, int k) {
    int i = 0;
    ListNode save = head, h, prev;
    while (i < k) {
      if (head == null) {
        return save;
      }
      head = head.next;
      i++;
    }
    i = 0;
    prev = head;
    head = save;
    while (i < k) {
      //TODO 头插法
      h = head.next;
      if (i == 0) {
        //根据情况递归
        head.next = reverseKGroup(prev, k);
      } else {
        head.next = prev;
      }
      prev = head;
      head = h;
      i++;
    }
    return prev;
  }
}
